overlord/configstate/configcore/swap.go

```diff
@@ -0,0 +1,137 @@
+	"path/filepath"
+	"time"
+
+	"github.com/mvo5/goconfigparser"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/systemd"
+)
+
+func init() {
+	supportedConfigurations["core.swap.size"] = true
+}
+
+func validateSystemSwapConfiguration(tr config.ConfGetter) error {
+	output, err := coreCfg(tr, "swap.size")
+	if err != nil {
+		return err
+	}
+
+	if output == "" {
+		return nil
+	}
+
+	// valid option for swap is any quantity unit that is larger than or equal
+	// to 1 MB
+	sz, err := quantity.ParseSize(output)
+	if sz < quantity.SizeMiB {
+		return fmt.Errorf("swap size of %s too small", output)
+	}
+	return err
+}
+
+func handlesystemSwapConfiguration(tr config.ConfGetter, opts *fsOnlyContext) error {
+	var pristineSwapSize, newSwapSize string
+	if err := tr.GetPristine("core", "swap.size", &pristineSwapSize); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if err := tr.Get("core", "swap.size", &newSwapSize); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if pristineSwapSize == newSwapSize {
+		return nil
+	}
+
+	// if it's unset, then treat it as if the size is "0" to not use swap by
+	// default
+	if newSwapSize == "" {
+		newSwapSize = "0"
+	}
+
+	szBytes, err := quantity.ParseSize(newSwapSize)
+	if err != nil {
+		return err
+	}
+
+	rootDir := dirs.GlobalRootDir
+	if opts != nil {
+		rootDir = opts.RootDir
+	}
+
+	swapConfigPath := filepath.Join(rootDir, "/etc/default/swapfile")
+
+	// TODO: also support writing/setting the location of the swap file setting?
+
+	// default location of the swapfile in case we can't determine the location
+	// from the config file
+	location := "/var/tmp/swapfile.swp"
+	if osutil.FileExists(swapConfigPath) {
+		// then get values from the config file
+		// read the existing file to get the location setting
+		cfg := goconfigparser.New()
+		cfg.AllowNoSectionHeader = true
+
+		if err := cfg.ReadFile(swapConfigPath); err != nil {
+			return err
+		}
+
+		location, err = cfg.Get("", "FILE")
+		if err != nil {
+			return err
+		}
+	}
+
+	// ensure the directory exists
+	if err := os.MkdirAll(filepath.Dir(swapConfigPath), 0755); err != nil {
+		return err
+	}
+
+	// the size of swap needs to be specified in Megabytes, while quantity.Size
+	// is a uint64 of bytes
+	fileContent := fmt.Sprintf("FILE=%s\nSIZE=%d\n", location, szBytes/quantity.SizeMiB)
+
+	// write the swap config file
+	if err := ioutil.WriteFile(swapConfigPath, []byte(fileContent), 0644); err != nil {
+		return err
+	}
+
+	if opts == nil {
+		// if we are not doing this filesystem only, then we need to also
+		// restart the swap service
+		sysd := systemd.NewUnderRoot(dirs.GlobalRootDir, systemd.SystemMode, &backlightSysdLogger{})
+
+		// TODO: what's an appropriate amount of time to wait here?
+		if err := sysd.Restart("swapfile.service", 60*time.Second); err != nil {
```

Comment:
on my UC20 this is a disabled service by default, does *.Restart() also enable it ?

Reply:
Well AFAIR, the service just calls swapon swapoff, etc. so it doesn't need to be running persistently. I did test it locally on UC20 and it works after a reboot, etc.

Reply:
yeah indeed it does, in a VM it took about 30 seconds to execute `snap set system swap.size=4G`